File: tests/r_parity/86_fect.R
# StatsPAI fect parity (R side) -- Module 86.
#
# Reads data/86_fect.csv (the staggered two-factor panel written by
# 86_fect.py) and runs fect::fect three times on the same bytes:
#   fe  : method = "fe"
#   ife : method = "ife", r = 2
#   mc  : method = "mc",  lambda = 0.002
# all with force = "two-way", se = FALSE, CV = FALSE, tol = 1e-12,
# max.iteration = 20000. Rows mirror the Python side:
#   <m>_att_avg, <m>_att_avg_unit, <m>_beta_x1, <m>_beta_x2, <m>_mu,
#   <m>_rmse, <m>_att_on_<k> (n = fect's cell count at relative period k;
#   fect's coding: 0 = last untreated period, 1 = first treated period).
# Tolerance: rel_est 1e-6 (deterministic EM fixed point on both sides).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(fect)
})

MODULE <- "86_fect"

# fect's initial two-way fit is a fixest::feols call with the package-wide
# demeaning tolerance (default 1e-6; the package floor is 2.2e-12). The Python port solves that initial
# least-squares problem exactly, and fect's one-pass "fe" model stops
# right after it, so the default tolerance would leave a ~1e-8 residual
# on the fe rows. Tightening fixest's own global option keeps the R side
# at the exact solution too.
fixest::setFixest_estimation(fixef.tol = 1e-11)

df <- read_csv_strict(MODULE)
n_obs <- nrow(df)

run_one <- function(method, ...) {
  fect::fect(
    Y ~ D + X1 + X2, data = df, index = c("id", "time"),
    method = method, force = "two-way", se = FALSE, CV = FALSE,
    tol = 1e-12, max.iteration = 20000, ...
  )
}

fits <- list(
  fe  = run_one("fe"),
  ife = run_one("ife", r = 2),
  mc  = run_one("mc", lambda = 0.002)
)

rows <- list()
extra <- list(
  force = "two-way", tol = 1e-12, max_iteration = 20000, r_ife = 2,
  lambda_mc = 0.002, fect_version = as.character(packageVersion("fect"))
)

for (m in names(fits)) {
  o <- fits[[m]]
  beta <- as.numeric(o$beta)
  add <- function(stat, est, n = n_obs) {
    rows[[length(rows) + 1L]] <<- parity_row(
      module = MODULE, statistic = stat, estimate = est, n = n
    )
  }
  add(paste0(m, "_att_avg"), o$att.avg)
  add(paste0(m, "_att_avg_unit"), o$att.avg.unit)
  add(paste0(m, "_beta_x1"), beta[1])
  add(paste0(m, "_beta_x2"), beta[2])
  add(paste0(m, "_mu"), o$mu)
  add(paste0(m, "_rmse"), o$rmse)
  for (j in seq_along(o$time)) {
    add(paste0(m, "_att_on_", o$time[j]), o$att[j], n = o$count[j])
  }
  extra[[paste0(m, "_niter")]] <- o$niter
  if (m == "mc") extra[["mc_lambda_norm"]] <- o$lambda.norm
}

write_results(MODULE, rows, extra = extra)

File: tests/r_parity/86_fect.py
"""StatsPAI fect counterfactual-estimator parity (Python side) -- Module 86.

Runs the native ``sp.fect`` port of Liu, Wang and Xu's counterfactual
estimators on a deterministic staggered-adoption factor panel and emits
one block of rows per outcome model:

- ``fe``  : two-way fixed effects (imputation estimator);
- ``ife`` : interactive fixed effects with ``r = 2`` factors, EM on the
  incomplete untreated panel;
- ``mc``  : matrix completion with a fixed nuclear-norm penalty.

Rows per method: ``<m>_att_avg`` (ATT over treated observations),
``<m>_att_avg_unit`` (mean of per-unit ATTs), ``<m>_beta_x1`` /
``<m>_beta_x2`` (covariate coefficients), ``<m>_mu`` (grand mean),
``<m>_rmse`` (pre-treatment fit on treated units), and
``<m>_att_on_<k>`` (ATT by relative period in fect's coding, where 0 is
the last untreated period and 1 the first treated period; ``n`` carries
fect's cell count). The companion 86_fect.R runs ``fect::fect`` with
``se = FALSE``, ``CV = FALSE``, ``force = "two-way"``, ``tol = 1e-12`` and ``max.iteration = 20000``
on the same CSV bytes; the Stata port (``fect_stata``, installed from
GitHub into a local ado path) runs the same three specifications.

Tolerance: rel_est 1e-6 (machine level; every row is a deterministic
fixed point of the same EM map started from the same two-way initial
fit). No SE rows: fect's inference is resampling-based and optional.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "86_fect"
R_FACTORS = 2
LAMBDA_MC = 0.002


def make_data(seed: int = PARITY_SEED) -> pd.DataFrame:
    """Staggered adoption with a two-factor error structure and covariates.

    60 units x 20 periods: 20 never-treated units and four cohorts of 10
    units first treated in periods 8, 11, 14, and 17 (so the earliest
    cohort has 7 untreated periods, above fect's ``min.T0 = 5`` for the
    ife/mc models). The treatment effect grows with time since
    treatment and varies by unit, and both factors load on treatment
    timing so a plain two-way model is biased -- which is what separates
    the three rows.
    """
    rng = np.random.default_rng(seed)
    N, T = 60, 20
    units = np.arange(1, N + 1)
    periods = np.arange(1, T + 1)
    first_treat = np.zeros(N, dtype=int)
    for c, start in enumerate((8, 11, 14, 17)):
        first_treat[20 + 10 * c : 30 + 10 * c] = start
    alpha = rng.normal(0.0, 1.0, N)
    xi = rng.normal(0.0, 1.0, T)
    F = rng.normal(0.0, 1.0, (T, R_FACTORS))
    L = rng.normal(0.0, 1.0, (N, R_FACTORS))
    # Loadings correlate with adoption timing (earlier cohorts load more).
    L[:, 0] += np.where(first_treat > 0, (20 - first_treat) / 6.0, 0.0)
    tau_unit = rng.normal(1.0, 0.3, N)
    rows = []
    for i, u in enumerate(units):
        for t_idx, t in enumerate(periods):
            x1 = rng.normal(0.0, 1.0) + 0.3 * L[i, 1]
            x2 = rng.normal(0.0, 1.0)
            d = int(first_treat[i] > 0 and t >= first_treat[i])
            since = (t - first_treat[i] + 1) if d else 0
            effect = tau_unit[i] * (1.0 + 0.2 * since) if d else 0.0
            y = (
                5.0 + alpha[i] + xi[t_idx] + F[t_idx] @ L[i]
                + 1.0 * x1 - 0.5 * x2 + effect + rng.normal(0.0, 0.5)
            )
            rows.append({"id": int(u), "time": int(t), "Y": float(y), "D": d,
                         "X1": float(x1), "X2": float(x2)})
    return pd.DataFrame(rows)


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    specs = {
        "fe": {},
        "ife": {"r": R_FACTORS},
        "mc": {"lam": LAMBDA_MC},
    }
    rows: list[ParityRecord] = []
    extra: dict = {
        "design": "60 units x 20 periods; 20 never-treated; cohorts first "
                  "treated in 8/11/14/17; two latent factors; two covariates",
        "force": "two-way",
        "tol": 1e-12,
        "max_iter": 20000,
        "r_ife": R_FACTORS,
        "lambda_mc": LAMBDA_MC,
        "relative_time_coding": (
            "fect: 0 = last untreated period, 1 = first treated period "
            "(StatsPAI relative_time = fect_time - 1)"
        ),
    }
    n_obs = int(len(df))
    for m, kw in specs.items():
        fit = sp.fect(
            df, y="Y", treat="D", unit="id", time="time",
            covariates=["X1", "X2"], method=m, tol=1e-12, max_iter=20000, **kw,
        )
        mi = fit.model_info
        rows.append(ParityRecord(module=MODULE, side="py", statistic=f"{m}_att_avg",
                                 estimate=float(fit.estimate), n=n_obs))
        rows.append(ParityRecord(module=MODULE, side="py", statistic=f"{m}_att_avg_unit",
                                 estimate=float(mi["att_avg_unit"]), n=n_obs))
        rows.append(ParityRecord(module=MODULE, side="py", statistic=f"{m}_beta_x1",
                                 estimate=float(mi["beta"]["X1"]), n=n_obs))
        rows.append(ParityRecord(module=MODULE, side="py", statistic=f"{m}_beta_x2",
                                 estimate=float(mi["beta"]["X2"]), n=n_obs))
        rows.append(ParityRecord(module=MODULE, side="py", statistic=f"{m}_mu",
                                 estimate=float(mi["mu"]), n=n_obs))
        rows.append(ParityRecord(module=MODULE, side="py", statistic=f"{m}_rmse",
                                 estimate=float(mi["pre_treatment_rmse"]), n=n_obs))
        for _, r in fit.detail.iterrows():
            rows.append(ParityRecord(
                module=MODULE, side="py",
                statistic=f"{m}_att_on_{int(r['fect_time'])}",
                estimate=float(r["att"]), n=int(r["count"]),
            ))
        extra[f"{m}_niter"] = int(mi["niter"])
        extra[f"{m}_converged"] = bool(mi["converged"])
        if m == "mc":
            extra["mc_lambda_norm"] = float(mi["lambda_norm"])

    write_results(MODULE, "py", rows, extra=extra)


if __name__ == "__main__":
    main()
